src/libraries/System.Linq/src/System/Linq/Cast.cs

```diff
@@ -55,5 +61,49 @@ private static IEnumerable<TResult> CastIterator<TResult>(IEnumerable source)
                 yield return (TResult)obj;
             }
         }
+
+        [DebuggerDisplay("Count = {Count}")]
+        private sealed partial class CastICollectionIterator<TResult>(ICollection source) : Iterator<TResult>
+        {
+            private readonly ICollection _source = source;
+            private IEnumerator? _enumerator;
+
+            public override Iterator<TResult> Clone() => new CastICollectionIterator<TResult>(_source);
+
+            public override bool MoveNext()
+            {
+                switch (_state)
+                {
+                    case 1:
+                        _enumerator = _source.GetEnumerator();
+                        _state = 2;
+                        goto case 2;
+
+                    case 2:
+                        Debug.Assert(_enumerator != null);
+                        if (_enumerator.MoveNext())
+                        {
+                            _current = (TResult)_enumerator.Current;
+                            return true;
+                        }
+
+                        Dispose();
+                        break;
+                }
+
+                return false;
+            }
+
+            public override void Dispose()
+            {
+                if (_enumerator != null)
+                {
+                    (_enumerator as IDisposable)?.Dispose();
+                    _enumerator = null;
+                }
+
+                base.Dispose();
```

Comment:
Nit: how about
```suggestion
                if (_enumerator is IDisposable d)
                {
                    d.Dispose();
                }
                
                _enumerator = null;
                base.Dispose();
```

Reply:
You mean just remove the outer if with the null check?
```C#
(_enumerator as IDisposable)?.Dispose();
```
and
```C#
if (_enumerator is IDisposable d)
{
    d.Dispose();
}
```
are the same thing.

Reply:
That works too.